### PR TITLE
[16.0][ADD] sale_stock_available_from_route_location

### PR DESCRIPTION
**Pull request by QuocDuong1306**
Currently, the forecasted and available quantities on the `Availability` widget of Sale Order are based on all locations of the warehouse
This module will re-calculate those quantities and base them on locations defined on `route_id` of Sale Order Line

### Notes for reviewers:
- Conflict with `stock-logistics-availability/pull/26` ( because `free_qty_today` based on `free_qty` of product) --> Not install both of them in `1 DB`

- In native Odoo, all the values of qty at date are calculated based on warehouse at [here](https://github.com/odoo/odoo/blob/a9c6d607ec81db83525074c298987e7a821d827b/addons/sale_stock/models/sale_order_line.py#L96)

- Tried many ways to extend the method to re-calculate based on locations, but the logic in it is very complicated and only supports warehouse.

- So I have overridden it completely and replaced the `warehouse calculation logic` with `location calculation logic` (should compare 2 versions for easier review)

### CODE REVIEW

File: sale_stock_route_location/models/sale_order_line.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        "product_uom_qty",
+        "product_uom",
+        "order_id.commitment_date",
+        "move_ids",
+        "move_ids.forecast_expected_date",
+        "move_ids.forecast_availability",
+        "route_id.rule_ids",
+    )
+    def _compute_qty_at_date(self):
+        # Override this method to re-calculate the availability quantities based on
```
Comment (nilshamerlinck) [inline sale_stock_route_location/models/sale_order_line.py]: include here a permalink to the original method that you are overriding

File: sale_stock_route_location/readme/DESCRIPTION.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Currently, the forecasted and available quantities on `Availability` widget of Sale Order are based on all locations of the warehouse
+This module will re-calculate those quantities and base on locations defined on `route_id` of Sale Order Line
```
Comment (nilshamerlinck) [inline sale_stock_route_location/readme/DESCRIPTION.rst]:
```suggestion
This module will re-calculate those quantities based only on the locations defined for the defined route of the Sale Order Line
```

File: sale_stock_route_location/readme/DESCRIPTION.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Currently, the forecasted and available quantities on `Availability` widget of Sale Order are based on all locations of the warehouse
```
Comment (nilshamerlinck) [inline sale_stock_route_location/readme/DESCRIPTION.rst]:
```suggestion
Currently, the forecasted and available quantities on `Availability` widget of Sale Order are based on all locations of the warehouse.
```

**Comment by nilshamerlinck** [comment]
instead of fully overriding `_compute_qty_at_date`, could you try to:
1) split it directly in odoo, and do a PR on your odoo fork
2) re-use some of these parts here

**Comment by QuocDuong1306** [comment]
Hi @nilshamerlinck , I updated the new fixup commit, could you have a look?

Depend:
- https://github.com/QuocDuong1306/odoo/pull/1

**Review by nilshamerlinck — APPROVED** [review]
pre-approving

can you please add a `readme/DEVELOP.rst` in which you explain 2 things:
- the implementation relies on the capacity of native odoo to calculate quantities based on a warehouse or a location passed in the context in `https://github.com/odoo/odoo/blob/055184ea033e7068ce33c92390c73ae39b2db259/addons/stock/models/product.py#L271`
- but it needs a patch on odoo upstream to generalize `_compute_qty_at_date`, see [YOUR PR]  

the same content should appear in your final oca PR description

**Pull request closed without merging**